gobblin-config-management/src/main/java/gobblin/config/client/ConfigClient.java

```diff
@@ -0,0 +1,257 @@
+package gobblin.config.client;
+
+import gobblin.config.configstore.ConfigStore;
+import gobblin.config.configstore.ConfigStoreFactory;
+import gobblin.config.configstore.ConfigStoreWithImportedBy;
+import gobblin.config.configstore.ConfigStoreWithImportedByRecursively;
+import gobblin.config.configstore.ConfigStoreWithResolution;
+import gobblin.config.configstore.ConfigStoreWithStableVersion;
+import gobblin.config.configstore.impl.ETLHdfsConfigStoreFactory;
+import gobblin.config.configstore.impl.SimpleConfigStoreResolver;
+import gobblin.config.configstore.impl.SimpleImportMappings;
+import gobblin.config.utils.URIComparator;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+import com.typesafe.config.Config;
+
+
+/**
+ * ConfigClient is the consumer facing class for retrieving configuration against all configuration stores.
+ * @author mitu
+ *
+ */
+public class ConfigClient {
+
+  public static enum VERSION_STABILITY_POLICY {
+    RAISE_ERROR,
+    CACHE_CONFIG_IN_MEMORY
+  }
+
+  private final VERSION_STABILITY_POLICY policy;
+
+  // key is the store ROOT, must use TreeMap
+  private final TreeMap<URI, ConfigStoreAccessor> configStoreMap = new TreeMap<URI, ConfigStoreAccessor>(
+      new URIComparator());
+
+  private ConfigClient(VERSION_STABILITY_POLICY policy) {
+    this.policy = policy;
+  }
+
+  /**
+   * @param policy - the policy control the behavior when this configuration client try to connect to configuration store.
+   * if policy is RAISE_ERROR and this client try to connect to any {@gobblin.config.configstore.ConfigStore} which is not
+   * {@gobblin.config.configstore.ConfigStoreWithStableVersion}, the RuntimeException will be thrown
+   * 
+   * if policy is CACHE_CONFIG_IN_MEMORY and this client try to connect to any {@gobblin.config.configstore.ConfigStore} which is not
+   * {@gobblin.config.configstore.ConfigStoreWithStableVersion}, the consumer need to cache the retrieved configuration in memory,
+   * otherwise, the same query for the same URI against same configuration store may result different result.
+   * @return - the Configuration Client object
+   */
+  public static ConfigClient createConfigClientWithStableVersion(VERSION_STABILITY_POLICY policy) {
+    return new ConfigClient(policy);
+  }
+
+  /**
+   * 
+   * @return the {@ConfigClient} with policy set to VERSION_STABILITY_POLICY.RAISE_ERROR
+   */
+  public static ConfigClient createDefaultConfigClient() {
+    // create with stable versions
+    return createConfigClientWithStableVersion(VERSION_STABILITY_POLICY.RAISE_ERROR);
+  }
+
+  private ConfigStoreAccessor getConfigStoreAccessor(URI uri) throws Exception {
+    if (this.configStoreMap.containsKey(uri)) {
```

Comment:
You want to check that the map contains an ancestor of uri, not necessarily the uri itself.